source/common/grpc/grpc_web_filter.cc

```diff
@@ -0,0 +1,128 @@
+    is_text_request_ = true;
+  }
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+
+  const Http::HeaderEntry* accept = headers.get(Http::Headers::get().Accept);
+  if (accept != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == accept->value().c_str()) {
+    // Checks whether gRPC-Web client is asking for b64 encoded response.
+    is_text_response_ = true;
+  }
+
+  // Adds te:trailers to upstream HTTP2 request. It's required for gRPC.
+  headers.addStatic(Http::Headers::get().TE, Http::Headers::get().TEValues.Trailers);
+  // Adds grpc-accept-encoding:identity,deflate,gzip. It's required for gRPC.
+  headers.addStatic(Http::Headers::get().GrpcAcceptEncoding,
+                    Http::Headers::get().GrpcAcceptEncodingValues.Default);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::decodeData(Buffer::Instance& data, bool) {
+  if (!is_text_request_) {
+    // No additional transcoding required if gRPC client is sending binary request.
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Parse application/grpc-web-text format.
+  if (data.length() + decoding_buffer_.length() < 4) {
+    decoding_buffer_.move(data);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t needed =
+      (data.length() + decoding_buffer_.length()) / 4 * 4 - decoding_buffer_.length();
+  decoding_buffer_.move(data, needed);
+  const std::string decoded = Base64::decode(
+      std::string(static_cast<const char*>(decoding_buffer_.linearize(decoding_buffer_.length())),
+                  decoding_buffer_.length()));
+  decoding_buffer_.drain(decoding_buffer_.length());
+  decoding_buffer_.move(data);
+  data.add(decoded);
+  return Http::FilterDataStatus::Continue;
+}
+
+// Implements StreamEncoderFilter.
+Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (is_text_response_) {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWebText);
+  } else {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+  if (!is_text_response_) {
+    // No additional transcoding required if gRPC-Web client asked for binary response.
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // The decoder always consumes and drains the given buffer. Incomplete data frame is buffered
+  // inside the decoder.
+  std::vector<Frame> frames;
+  decoder_.decode(data, frames);
+  if (frames.empty()) {
+    // We don't have enough data to decode for one single frame, stop iteration until more data
+    // comes in.
+    return Http::FilterDataStatus::StopIterationNoBuffer;
+  }
+
+  // Encodes the decoded frames with base64.
+  for (auto& frame : frames) {
+    Buffer::OwnedImpl temp;
+    temp.add(&frame.flags_, 1);
```

Comment:
It's gRPC frame, comes from backend. It's documented in the decoder.h.
https://github.com/lyft/envoy/blob/master/source/common/grpc/codec.h#L49